src/core/QCheck_ng.ml

```diff
@@ -0,0 +1,2052 @@
+    (* p_shrink1: ('a, 'b) t Iter.t;
+       p_shrink2: 'b Shrink.t -> ('a, 'b) t Iter.t; *)
+    p_print: unit -> string;
+  }
+
+  (** TODO This is a bit trickier because the function entries are populated at runtime
+      while the shrinking tree must be provided statically and is not mutable ([Seq.t]),
+      only the inside tables are.
+
+      We could arguably create a shrinking sequence of fixed size (where the first element is
+      a function with no entries) and spread in best-effort entries. Not doing it for now. *)
+  let create (type k) (type v) (k : k Observable.t) (v : v arbitrary) (size : int) : (k, v) t Gen.t =
+    fun st ->
+    let module T = Hashtbl.Make(struct
+        type t = k
+        let equal = k.Observable.eq
+        let hash = k.Observable.hash
+      end) in
+    (* let tbl_to_list tbl =
+     *   T.fold (fun k v l -> (k,v) :: l) tbl []
+     * and tbl_of_list l =
+     *   let tbl = T.create (max (List.length l) 8) in
+     *   List.iter (fun (k,v) -> T.add tbl k v) l;
+     *   tbl
+     * in *)
+    (* make a table
+       @param extend if true, extend table on the fly *)
+    let (* rec *) make ~extend tbl = {
+      get = (fun x ->
+          try Some (T.find tbl x)
+          with Not_found ->
+            if extend then (
+              let (Tree.Tree (v, _shrinks)) = v.gen st in
+              T.add tbl x v;
+              Some v
+            ) else None);
+      p_print = (fun () -> match v.print with
+          | None -> "<fun>"
+          | Some pp_v ->
+            let b = Buffer.create 64 in
+            T.iter
+              (fun key value ->
+                 Printf.bprintf b "%s -> %s; "
+                   (k.Observable.print key) (pp_v value))
+              tbl;
+            Buffer.contents b);
+      (* p_shrink1=(fun yield ->
+         Shrink.list (tbl_to_list tbl)
+          (fun l ->
+             yield (make ~extend:false (tbl_of_list l)))
+         );
+         p_shrink2=(fun shrink_val yield ->
+         (* shrink bindings one by one *)
+         T.iter
+          (fun x y ->
+             shrink_val y
+               (fun y' ->
+                  let tbl' = T.copy tbl in
+                  T.replace tbl' x y';
+                  yield (make ~extend:false tbl')))
+          tbl); *)
+      p_size=(fun size_v -> T.fold (fun _ v n -> n + size_v v) tbl 0);
+    } in
+    Tree.pure (make ~extend:true (T.create size))
+
+  let get t x = t.get x
+  (* let shrink1 t = t.p_shrink1
+     let shrink2 p t = t.p_shrink2 p *)
+  let print t = t.p_print ()
+  let size p t = t.p_size p
+end
+
+(** Internal representation of functions *)
+type ('a, 'b) fun_repr_tbl = {
+  fun_tbl: ('a, 'b) Poly_tbl.t;
+  fun_arb: 'b arbitrary;
+  fun_default: 'b;
+}
+
+type 'f fun_repr =
+  | Fun_tbl : ('a, 'ret) fun_repr_tbl -> ('a -> 'ret) fun_repr
+  | Fun_map : ('f1 -> 'f2) * 'f1 fun_repr -> 'f2 fun_repr
+
+type _ fun_ =
+  | Fun : 'f fun_repr * 'f -> 'f fun_
+
+(** Reifying functions *)
+module Fn = struct
```

Comment:
@c-cube about function generation and function showing/shrinking in general:
if it's ok with you, I propose to **not** implement function shrinking **in this MR** (function generation should still work) (or we can have a dumb shrinker, like "shrink to the constant function that returns the default value"), and instead create a separate issue to rework the function support (module `Fn`).

The reason for this is that the current implementation of function generation and shrinking uses Hash table mutation at test runtime to populate the table with entries, which doesn't compose well with rosetree shrinking.

Even if we forget about integrated shrinking, I have mixed feelings about this design because it means the generated function depends on runtime arguments passed to it: the same test run with the same seed would yield different (input) functions if the argument passed to it varies outside of the seed/random state (e.g. for PBTs that want to test a random function).

In particular I see there is a paper [Shrinking and showing functions: (functional pearl)](https://dl.acm.org/doi/abs/10.1145/2430532.2364516) which seems to address exactly that, and e.g. it was used as basis to implement [hedgehog-fn](https://hackage.haskell.org/package/hedgehog-fn) Haskell package, so this may be a better ground to implement the `Fn` module of QCheck.

What do you think?

Reply:
I haven't looked at the paper, but if you can do sth cleaner in the future, by all means, yes!

Reply:
Ah indeed it's paywalled :cry: 

I found this Youtube talk by the paper author Koen Claessen: https://www.youtube.com/watch?v=CH8UQJiv9Q4

In particular at 08:35 he explains an early first approach which looks exactly like the current implementation in QCheck, using mutation to record all called arguments, and explains some shortcomings.

The actual implementation, from what I understood, is not so far from the current QCheck one, except there is no mutation involved:
* only build partial functions (this is similar to `Poly_tbl` in QCheck) called `a :-> b`
* transform them into total functions by using a default value (similar to what is done in `fun_repr_tbl.fun_default` in QCheck)
* display and shrinking are done on `(a :-> b, b)` (so in fine, a `a b Fun.t` coulde be seen as:
```ocaml
module Fun = struct
  type 'a 'b Fun.t = Fun of (a :-> b, b) * (a -> b)
end
```
The test would just ignore the first argument and only extract the second (the "real" function) while showing and shrinking would only use the first (partial finite function which can be seen as as assoc list `('a * 'b) list` + default value).

Reply:
Right, but I don't really understand how the "relevant" keys are found in the first place. Seems very hit-or-miss if we just generate a random function without considering what it's applied to. :)

I think this can wait anyway, splittable generators and sized generators seem more important to me.